core/chains/evm/logpoller/log_poller.go

```diff
@@ -1121,13 +1125,35 @@ func (lp *logPoller) PruneOldBlocks(ctx context.Context) (bool, error) {
 	return lp.logPrunePageSize == 0 || rowsRemoved < lp.logPrunePageSize, err
 }
 
-// PruneExpiredLogs logs that are older than their retention period defined in Filter.
-// Returns whether all logs eligible for pruning were removed. If logPrunePageSize is set to 0, it will always return true.
+// PruneExpiredLogs will attempt to remove any logs which have passed their retention period. Returns whether all expired
+// logs were removed. If logPrunePageSize is set to 0, it will always return true unless an actual error is encountered
 func (lp *logPoller) PruneExpiredLogs(ctx context.Context) (bool, error) {
+	done := true
+
 	rowsRemoved, err := lp.orm.DeleteExpiredLogs(ctx, lp.logPrunePageSize)
-	return lp.logPrunePageSize == 0 || rowsRemoved < lp.logPrunePageSize, err
+	if err != nil {
+		lp.lggr.Errorw("Unable to find excess logs for pruning", "err", err)
+		return false, err
+	} else if lp.logPrunePageSize != 0 && rowsRemoved == lp.logPrunePageSize {
+		done = false
+	}
+
+	rowIDs, err := lp.orm.SelectExcessLogIDs(ctx, lp.logPrunePageSize)
```

Comment:
Should we have a feature flag to disable it if it causes issues in production?

Reply:
I took your other suggestion of not calling `PruneExcessLogs()` unless there is a filter registered with `MaxLogsKept > 0`.  That may be good enough for now, since there aren't even any products yet which set `MaxLogsKept > 0`

When this gets merged into the ccip repo and they are ready to start passing non-zero values for it, they can decide whether to make them configurable, hide them behind a feature flag, or just always hard code them to non-zero.  If it's the last they opt for, we can revisit adding a global feature flag for it, but having things like Retention and MaxLogsKept be configurable per product seems like something we ought to encourage, rather than an all-or-nothing approach.  For example, they could have a feature flag which sets either the Retention field or MaxLogsKept... but having both at once seems unnecessary and having only a MaxLogsKept that can be turned on or off in LogPoller's global configuration would mean turning it off goes back to "retain everything" rather than a time-based retention.  Curious to hear what @mateusz-sekara thinks though.

Reply:
I think enabling/disabling that with a feature flag adds another layer of complexity to understand of what is enabled and what not (and would probably make code complex as well). I would refrain from using it unless it's necessary. As @reductionista mentioned, this feature is not used by anyone else and probably we will be the first customer of that. This was already tested pretty heavily by @reductionista, but we will let it soak on pre-prod environment for some time before going to mainnet. That should give us a lot insights on what's the impact and how the query behaves in the long run.

In case of issues on prod-mainnet we can disable that by changing how the filter is configured instead of introducing another feature flag (bring back MaxLogsKept to the previous value). Does it make sense?